pydantic_ai_slim/pydantic_ai/ui/_web/api.py

```diff
@@ -0,0 +1,169 @@
+    """Extra data extracted from chat request."""
+
+    model: str | None = None
+    """Model ID selected by the user, e.g. 'openai:gpt-5'. Maps to JSON field 'model'."""
+    builtin_tools: list[str] = []
+    """Tool IDs selected by the user, e.g. ['web_search', 'code_execution']. Maps to JSON field 'builtinTools'."""
+
+
+def validate_request_options(
+    extra_data: ChatRequestExtra,
+    model_ids: set[str],
+    allowed_tool_ids: set[str],
+) -> str | None:
+    """Validate that requested model and tools are in the allowed lists.
+
+    Returns an error message if validation fails, or None if valid.
+    """
+    if extra_data.model and model_ids and extra_data.model not in model_ids:
+        return f'Model "{extra_data.model}" is not in the allowed models list'
+
+    # base model also validates this but makes sense to have an api check, since one could be a UI bug/misbehavior
+    # the other would be a pydantic-ai bug
+    # also as future proofing since we don't know how users will use this feature in the future
+    invalid_tools = [t for t in extra_data.builtin_tools if t not in allowed_tool_ids]
+    if invalid_tools:
+        return f'Builtin tool(s) {invalid_tools} not in the allowed tools list'
+
+    return None
+
+
+def create_api_routes(
+    agent: Agent[AgentDepsT, OutputDataT],
+    models: ModelsParam = None,
+    builtin_tools: Sequence[AbstractBuiltinTool] | None = None,
+    deps: AgentDepsT = None,
+    model_settings: ModelSettings | None = None,
+    instructions: str | None = None,
+) -> list[Route]:
+    """Create API routes for the web chat UI.
+
+    Args:
+        agent: Agent instance.
+        models: Models to make available in the UI. Can be:
+            - A sequence of model names/instances (e.g., `['openai:gpt-5', Model(...)]`)
+            - A dict mapping display labels to model names/instances
+            If not provided, the UI will have no model options.
+        builtin_tools: Optional list of additional builtin tools to make available in the UI.
+            Tools already configured on the agent are always included but won't appear as options.
+        deps: Optional dependencies to use for all requests.
+        model_settings: Optional settings to use for all model requests.
+        instructions: Optional extra instructions to pass to each agent run.
+
+    Returns:
+        A list of Starlette Route objects for the API endpoints.
+    """
+    # Build model ID → original reference mapping and ModelInfo list for frontend
+    model_id_to_ref: dict[str, Model | str] = {}
+    model_infos: list[ModelInfo] = []
+
+    # Filter out builtin_tools that are already configured on the agent (they're always included)
+    agent_tool_types = {type(t) for t in agent._builtin_tools if isinstance(t, AbstractBuiltinTool)}  # pyright: ignore[reportPrivateUsage]
+    ui_builtin_tools = [t for t in (builtin_tools or []) if type(t) not in agent_tool_types]
```

Comment:
This breaks if there is a single `MCPServerTool` in `agent._builtin_tools` and a new oen is passed in via `builtin_tools`, as they both have the same type. We have to use `unique_id` for things like this!!!